node/network/collator-protocol/src/validator_side/mod.rs

```diff
@@ -1374,45 +1620,67 @@ async fn handle_collation_fetched_result<Context>(
 		},
 	};
 
-	if let Some(collations) = state.collations_per_relay_parent.get_mut(&relay_parent) {
-		if let CollationStatus::Seconded = collations.status {
-			gum::debug!(
-				target: LOG_TARGET,
-				?relay_parent,
-				"Already seconded - no longer interested in collation fetch result."
-			);
-			return
-		}
-		collations.status = CollationStatus::WaitingOnValidation;
-	}
+	let collations = &mut per_relay_parent.collations;
+	// There's always a single collation being fetched at any moment of time.
+	// In case of a failure, we reset the status back to waiting.
+	collations.status = CollationStatus::WaitingOnValidation;
 
-	if let Entry::Vacant(entry) = state.pending_candidates.entry(relay_parent) {
+	let fetched_collation = FetchedCollation::from(&candidate_receipt);
+	if let Entry::Vacant(entry) = state.fetched_candidates.entry(fetched_collation) {
 		collation_event.1.commitments_hash = Some(candidate_receipt.commitments_hash);
 
-		if let Some(pvd) = request_persisted_validation_data(
-			ctx,
-			candidate_receipt.descriptor().relay_parent,
-			candidate_receipt.descriptor().para_id,
-		)
-		.await
-		{
-			// TODO [https://github.com/paritytech/polkadot/issues/5054]
-			//
-			// If PVD isn't available (core occupied) then we'll silently
-			// just not second this. But prior to asynchronous backing
-			// we wouldn't second anyway because the core is occupied.
-			//
-			// The proper refactoring would be to accept declares from collators
-			// but not even fetch from them if the core is occupied. Given 5054,
-			// there's no reason to do this right now.
-			ctx.send_message(CandidateBackingMessage::Second(
```

Comment:
Are there checks here to ensure we only second one per depth per active-leaf? We should also be issuing `Second` messages when active-leaves change, for candidates we haven't tried to second yet.

Reply:
It is true that candidate-backing will refuse to second anything outside of those limitations already. But if we are punting those concerns onto candidate-backing, we should document that.

Reply:
(was accidentally reviewing the old diff)